src/deploy/index.ts

```diff
@@ -61,7 +63,12 @@ export const deploy = async function (
     const config = options.config.get("hosting");
     if (Array.isArray(config) ? config.some((it) => it.source) : config.source) {
       experiments.assertEnabled("webframeworks", "deploy a web framework to hosting");
+      const usedToTargetFunctions = targetNames.includes("functions");
       await prepareFrameworks(targetNames, context, options);
+      const nowTargetsFunctions = targetNames.includes("functions");
+      if (nowTargetsFunctions && !usedToTargetFunctions) {
+        await requirePermissions(TARGET_PERMISSIONS["functions"]);
```

Comment:
This is going to break every GH action deployment using webframeworks because they don't have the permissions to deploy functions (such is my understanding). We should have a better story of how to upgrade those users' SAs before we push this change.

Reply:
Then again, if they're broken _now_, that's unlikely to matter.

Reply:
We need to focus on a sprint to give customers preview channels for web frameworks + pinned tags. I can play with the code, but I don't know how to test outside of production.